### PR TITLE
Remove __cplusplus preamble from internal headers

### PR DESCRIPTION
**Pull request by romen**
From https://github.com/openssl/openssl/pull/6521#discussion_r197262781 , spotted by @dot-asm :

> This is internal header and it's never exposed to C++ compiler, hence there is no need for this. 

This was later extended to all the internal headers I could think of!

### CODE REVIEW

**Comment by dot-asm** [comment]
How about a request what would omit cplusplus from **all** internal headers? I mean why chase it one-by-one...

**Comment by mspncp** [comment]
> How about a request what would omit cplusplus from all internal headers?

Sounds like a good idea. @romen, are you willing to extend your pull request? If you are too busy with all the other pull requests, then I will volunteer to do it.

**Comment by romen** [comment]
No problem!

**Comment by romen** [comment]
@dot-asm @mspncp @mattcaswell this should be ready now, unless you can point me to other internal headers I did not think of: I checked `{crypto,ssl}/**/*.h` and `include/internal/*.h`.

(this should be *autosquashed* before merging)

File: crypto/hmac/hmac_lcl.h [hunk 1]
```diff
@@ -10,13 +10,6 @@
 #ifndef HEADER_HMAC_LCL_H
 # define HEADER_HMAC_LCL_H
 
-#ifdef  __cplusplus
-extern "C" {
-#endif
-#if 0                            /* emacs indentation fix */
-}
-#endif
-
```
Comment (romen) [inline crypto/hmac/hmac_lcl.h]: What was this? I went ahead and removed it, should I check if there are other similar emacs specific tricks in the code??

**Comment by mspncp** [comment]
I checked the tip of your pr using the following search. This should find all locations except for public headers and external libraries.

```
msp@msppc:~/src/openssl$ find * -name '*.h' -exec grep -l __cplusplus {} + | grep -v -e include/openssl -e krb5 -e boringssl
e_os.h
```
The search found one more location in `e_os.h`.

File: crypto/hmac/hmac_lcl.h (hunk 1 above)
Comment (levitte) [reply crypto/hmac/hmac_lcl.h]: I had forgotten about those. That's so emacs and other indentation programs don't add extra spaces on everything that's between

``` C
extern "C" {
```

and the corresponding

``` C
} 
```

We should actually use this trick anywhere we have these C++ lines, that would save some cuss words...

Comment (levitte) [reply crypto/hmac/hmac_lcl.h]: It's fine to remove that trick when you remove the C++ stuff, of course

Comment (mspncp) [reply crypto/hmac/hmac_lcl.h]: I think it's the only location.

Comment (romen) [reply crypto/hmac/hmac_lcl.h]: My Vim is different! 😝

Comment (mspncp) [reply crypto/hmac/hmac_lcl.h]: No vim-vs-emacs flames, please ;-)

**Comment by mspncp** [comment]
I think this can be merged as soon as the ci bots are happy.

**Comment by romen** [comment]
@mspncp thanks for the convenient oneliner!
Latest commit fixes `e_os.h`.

File: crypto/hmac/hmac_lcl.h (hunk 1 above)
Comment (romen) [reply crypto/hmac/hmac_lcl.h]: > No vim-vs-emacs flames, please ;-)

Oh no, I was already making popcorn!!

Comment (levitte) [reply crypto/hmac/hmac_lcl.h]: #6557 means we won't need that kind of trick any more.

**Comment by levitte** [comment]
Side note: 4 different people approving!  I think this is a record number ;-)

**Comment by levitte** [comment]
Oh, I forgot about that one!

**Comment by mspncp** [comment]
I didn't. It was my first big pull request :-).

**Comment by mspncp** [comment]
Done! Thanks @romen for doing the cleanup!

**Pull request closed without merging**